das/daser.go

```diff
@@ -75,6 +116,57 @@ func (d *DASer) sampling(ctx context.Context, sub header.Subscription) {
 			continue
 		}
 
+		// If the next header coming through gossipsub is not adjacent
+		// to our last DASed header, kick off routine to DAS all headers
+		// between last DASed header and h. This situation could occur
+		// either on start or due to network latency/disconnection.
+		if h.Height > (checkpoint + 1) {
+			// DAS headers between last DASed height up to the current
+			// header
+			go d.catchUp(ctx, checkpoint, h.Height-1)
+		}
+
+		startTime := time.Now()
+
+		err = d.da.SharesAvailable(ctx, h.DAH)
+		if err != nil {
+			if err == context.Canceled {
+				return
+			}
+			log.Errorw("sampling failed", "height", h.Height, "hash", h.Hash(),
+				"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "err", err)
+			// continue sampling
+		}
+
+		sampleTime := time.Since(startTime)
+		log.Infow("sampling successful", "height", h.Height, "hash", h.Hash(),
+			"square width", len(h.DAH.RowsRoots), "finished (s)", sampleTime.Seconds())
+
+		checkpoint = h.Height
+	}
+}
+
+// catchUp starts a sampling routine for headers starting at the `from`
+// height and exits the loop once `to` is reached. (from:to]
+func (d *DASer) catchUp(ctx context.Context, from, to int64) {
+	defer close(d.catchUpDn)
```

Comment:
There can be multiple `catchUp` routines during runtime, so this will panic if e.g. I close my laptop lid twice in a row for more than block-time time. 

I see here two solutions:
* Making a `catchUp` struct with its own lifecycle, so `sample` will be able to create `catchUp` many instances and wait until they are all done on `defer`
* Spawning one only `catchUp` routine that communicates with `sample` by a chan, i.e. `sample` sends to `catchUp` structs so the routine process them in the order.

I believe the second approach is better, as it simplifies parallelization and accounting for in-progress catchups to be provided with DASer API. The parallelization can be done with `catchUpWorker`s controlled by the `catchUp` routine, so once the job is rcvd from the `sampler`, `catchUp` routines equally splits according to then N of workers and pushes to another chan that the workers listen to.
The `catchUp` routine would also control the accounting of heights being DASed or `State()`

Reply:
resolved by https://github.com/celestiaorg/celestia-node/pull/473/commits/55d198fe32faf83acab201a5325ddb15a93ba018 ?